WordPress/gutenberg
# [RNMobile] Update "add block" button style in default editor view

Fixes: https://github.com/wordpress-mobile/gutenberg-mobile/issues/4692

## Related PRs and Installable Builds

* `Gutenberg Mobile:` https://github.com/wordpress-mobile/gutenberg-mobile/pull/4705
* `Android installable build:` https://github.com/wordpress-mobile/WordPress-Android/pull/16196
* `iOS installable build:` https://github.com/wordpress-mobile/WordPress-iOS/pull/18233

## What?

This PR seeks to make the block inserter more visible in the "default view" of the editor. For the purposes of this PR, "default view" has been interpreted as the view of the editor when no blocks (including the editor's title) are selected.

## Why?

User research indicates that it's not always clear to users what primary actions are available when they first open the editor. By iterating on the default view of the editor, it's hoped that it will be easier for users to discover blocks and, therefore, improve the overall writing/onboarding experience.

## How?

With the changes from this PR, whenever a block _isn't_ selected the block inserter button will be styled more prominently and with an `Add Blocks` label. The following is a high-level overview of the code changes involved:

### Logic around "isDefaultView" check

A [new `noContentSelected` check](https://github.com/WordPress/gutenberg/blob/ee541300bae7268c34e359b0a7551fe5f2d0493f/packages/edit-post/src/components/header/header-toolbar/index.native.js#L178) has been implemented to determine whether a block or title is currently selected or not. If nothing is selected within the editor, then `noContentSelected` returns as `true`.

### Styling
 
The `noContentSelected` check is utilised to add different styles to the toolbar and block inserter button in cases when it returns as `true`:

* The `header-toolbar` component (the toolbar at the bottom of the editor) has been updated to [increase in height]([header-toolbar](https://github.com/WordPress/gutenberg/tree/82eeb5f8aeeca855849649479ea488f955d5cb45/packages/edit-post/src/components/header/header-toolbar)[header-toolbar](https://github.com/WordPress/gutenberg/tree/82eeb5f8aeeca855849649479ea488f955d5cb45/packages/edit-post/src/components/header/header-toolbar)) when `noContentSelected` is `true`. To enable this, a new `$mobile-header-toolbar-expanded-height` variable [has been added](https://github.com/WordPress/gutenberg/blob/ee541300bae7268c34e359b0a7551fe5f2d0493f/packages/base-styles/_variables.scss#L114) to the global `_variables.scss` file.
* The `inserter` component (the `+` button within the toolbar) has also been updated to [display a text label](https://github.com/WordPress/gutenberg/blob/ee541300bae7268c34e359b0a7551fe5f2d0493f/packages/block-editor/src/components/inserter/index.native.js#L55) and [apply new styles](https://github.com/WordPress/gutenberg/blob/ee541300bae7268c34e359b0a7551fe5f2d0493f/packages/block-editor/src/components/inserter/index.native.js#L48) when `noContentSelected` is `true`.

## Testing Instructions

To view and test the new button's new styles, follow these steps within the WordPress Android or iOS app:

* Open the editor and tap the button. 
* By default the title will be selected (this will be changed in a separate PR). In order to view the new button, close the keyboard so that no area of the editor is selected (block or title). 
* The `+` inserter will take up a bigger amount of the toolbar than previously, and will display `Add Blocks` text alongside it.
* Selecting a block within the editor will revert the inserter back to its previous style. 

As this is a UI change, it'd be helpful to test the following as part of the above flow:

* Appearance on dark mode vs. light mode.
* Animation when the inserter button changes from the default style and back.
* Display on different device sizes, including the portrait and landscape modes.
* Appearance with different translations, including longer translations and RTL languages. _Note, no translations have been added as of yet, so testing will need to be done by manually editing code with a local setup._
* Verify the text that's read by VoiceOver (iOS) or Talkback (Android) when the new expanded button style is shown is clear (with no regressions).

## Screenshots or screencast 

### General

These before/after screenshots show the inserter button (the `+` icon to the left of the bottom toolbar) as it displayed before the changes in this PR and after. Note, the styling in the "after" screenshot will only be visible when no part of the editor is selected (no block or title selection):

| Before  | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/162632365-13b35a32-a410-466c-ba15-33a82e0a4947.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/162632357-00a26760-9f4f-4abc-8efc-5744750af58b.png" width="100%"> |

### Dark mode

| Dark Mode - Default View | Dark Mode - Block View |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/162632027-20b8a9da-a0b4-4170-9ee5-bc78fe7e5e4c.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/162632062-b26c9712-7d0b-4d77-8910-4acd53233cca.png" width="100%"> |
